torch/_inductor/autotuner/model.py

```diff
@@ -0,0 +1,649 @@
+            return val > self.get_xmax()
+        if name == "YBLOCK":
+            return val > self.get_ymax()
+        if name == "ZBLOCK":
+            return val > self.get_zmax()
+        if name == "RBLOCK":
+            return val > self.get_rmax()
+
+        return False
+
+    def get_neighbour_values(self, name, orig_val, radius=1, include_self=False):
+        """
+        Get neighbour values in 'radius' steps. The original value is not
+        returned as it's own neighbour.
+        """
+        assert radius >= 1
+
+        def update(cur_val, inc=True):
+            if name == "num_stages":
+                if inc:
+                    return cur_val + 1
+                else:
+                    return cur_val - 1
+            else:
+                if inc:
+                    return cur_val * 2
+                else:
+                    return cur_val // 2
+
+        out = []
+        # increment loop
+        cur_val = orig_val
+        for _ in range(radius):
+            cur_val = update(cur_val, True)
+            if self.value_too_large(name, cur_val):
+                break
+            out.append(cur_val)
+
+        # decrement loop
+        cur_val = orig_val
+        for _ in range(radius):
+            cur_val = update(cur_val, False)
+            if cur_val <= 0:
+                break
+            out.append(cur_val)
+
+        if include_self:
+            out.append(orig_val)
+        return out
+
+    def get_field(self, config, name):
+        if name == "num_warps":
+            return config.num_warps
+        elif name == "num_stages":
+            return config.num_stages
+        else:
+            return config.kwargs.get(name, None)
+
+    def set_field(self, config, name, value):
+        if name == "num_warps":
+            config.num_warps = value
+        elif name == "num_stages":
+            config.num_stages = value
+        else:
+            config.kwargs[name] = value
+
+    def check_all_tuning_directions(self, st_config, radius):
+        candidate_values_list = []
+        effective_fields = []
+        for field in self.tunable_fields:
+            old_value = self.get_field(st_config, field)
+            if old_value is None:
+                continue
+            candidate_values = self.get_neighbour_values(
+                field,
+                old_value,
+                radius=radius,
+                include_self=True,
+            )
+            candidate_values_list.append(candidate_values)
+            effective_fields.append(field)
+
+        res = list()
+        choices = itertools.product(*candidate_values_list)
+        for choice in choices:
+            assert len(choice) == len(effective_fields)
+            candidate_config = copy.deepcopy(st_config)
+            for new_val, field in zip(choice, effective_fields):
+                self.set_field(candidate_config, field, new_val)
+            res.append(candidate_config)
+        return res
```

Comment:
Is this duplicate code from the coordinate descent tuner? Can we refactor so the search algorithm is shared?